File: leetCode/medium/17-letterCombinationsPhoneNumber.ts
function letterCombinationsPhoneNumber(digits: string) {
  if (digits === "") return [];

  const t9Map = new Map([
    ["1", ""],
    ["2", "abc"],
    ["3", "def"],
    ["4", "ghi"],
    ["5", "jkl"],
    ["6", "mno"],
    ["7", "pqrs"],
    ["8", "tuv"],
    ["9", "wxyz"],
    ["0", " "],
  ]);

  const result: string[] = [];

  const helper = (index: number, str: string) => {
    if (index === digits.length) {
      result.push(str);
      return;
    }

    const letters = t9Map.get(digits[index]) ?? "0";
    for (let i = 0; i < letters.length; i += 1) {
      helper(index + 1, str + letters[i]);
    }
  };

  helper(0, "");

  return result;
}

console.log(letterCombinationsPhoneNumber("23"));
console.log(letterCombinationsPhoneNumber("234"));
console.log(letterCombinationsPhoneNumber("56435"));

/** 
 000
 001
 002
 010
 011
 012
 020
 021
 022 
 */

File: leetCode/medium/17-letterCombinationsPhoneNumber.jl
# need to fix, fails some tests

function letterCombinationsPhoneNumber(digits::String)::Array{String,1}
  if length(digits) === 0
    return []
  end

  t9Map = Dict{String,String}(
    "1" => "",
    "2" => "abc",
    "3" => "def",
    "4" => "ghi",
    "5" => "jkl",
    "6" => "mno",
    "7" => "pqrs",
    "8" => "tuv",
    "9" => "wxyz",
    "0" => " "
  )

  results::Array{String,1} = []

  function helper(index::Int64, str::String)
    if index === length(digits)
      push!(results, str)
      return
    end

    letters = get(t9Map, digits[index], 1)
    for (idx, _) in pairs(letters)
      helper(index + 1, str * string(letters[idx]))
    end
  end

  helper(1, "")

  return results
end

using Test

@testset "letterCombinationsPhoneNumber" begin
  @test letterCombinationsPhoneNumber("23") == ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"]
  @test letterCombinationsPhoneNumber("2") == ["a", "b", "c"]
  @test letterCombinationsPhoneNumber("0") == [" "]
  @test letterCombinationsPhoneNumber("") == []
end

